### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by mkhapisov**
Adding MTCNN aligner and resnet18 recognizer

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,15 @@
-Solving face recognition task with this labeled train dataset: https://www.kaggle.com/datasets/jessicali9530/lfw-dataset/data
\ No newline at end of file
+Solving face recognition task using Pytorch, MTCNN (https://github.com/ipazc/mtcnn) as an aligner and resnet18 as a recognizer (classifier). 
+Used dataset: https://www.kaggle.com/datasets/jessicali9530/lfw-dataset/data
+
+Project's structure:
```
Comment (vBykoff) [inline README.md:8]: \```
Project's structure:
...

\```

File: README.md [hunk 2]
```diff
@@ -1 +1,15 @@
-Solving face recognition task with this labeled train dataset: https://www.kaggle.com/datasets/jessicali9530/lfw-dataset/data
\ No newline at end of file
+Solving face recognition task using Pytorch, MTCNN (https://github.com/ipazc/mtcnn) as an aligner and resnet18 as a recognizer (classifier).
```
Comment (vBykoff) [inline README.md]: Заголовок через #

File: Face Recognition/recognizer.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        return probs
+    
+    def predict_one_sample(self, sample):
+        with torch.no_grad():
+            self.model.eval()
+            sample = sample.to(self.device)
+            logit = self.model(sample).cpu()
+            probs = F.softmax(logit, dim=-1).numpy()
+        return probs
+
```
Comment (vBykoff) [inline Face Recognition/recognizer.py]: Можешь отформатировать код, установи black через pip и прогони свои файлы

Comment (vBykoff) [reply Face Recognition/recognizer.py]: он уберет тишние пробелы и все такое

File: Face Recognition/recognizer.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                print("Decreasing learning rate: %f -> %f" % (self.optimizer.param_groups[0]['lr'], self.optimizer.param_groups[0]['lr'] * self.lr_coef))
+                self.optimizer.param_groups[0]['lr'] *= self.lr_coef
+
+        print("Max score: %f, Min loss: %f, Epoch: %d" % (max_score, min_loss, epoch_max_score))
+        if epoch_max_score > -1:
+            checkpoint = torch.load('best_recognizer.pth')
+            self.model.load_state_dict(checkpoint['model_state_dict'])
+        return history
+    
+    def predict(self, test_loader):
```
Comment (vBykoff) [inline Face Recognition/recognizer.py]: не указан тип у test_loader

File: Face Recognition/recognizer.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        for inputs, labels in test_loader:
+            inputs = torch.FloatTensor(inputs).to(self.device)
+            with torch.no_grad():
+                outputs = self.model(inputs).cpu()
+                logits.append(outputs)
+        
+        probs = F.softmax(torch.cat(logits), dim=-1).numpy()
+        return probs
+    
+    def predict_one_sample(self, sample):
```
Comment (vBykoff) [inline Face Recognition/recognizer.py]: не указан тип sample и  типо возвращаемой переменной:     
```def predict_one_sample(self, sample:array) -> np,.array:```

File: Face Recognition/recognizer.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        self.optimizer = optimizer
+
+        self.lr_step = lr_step
+        self.lr_coef = lr_coef
+    
+    def freeze_layers(self, layers_to_unfreeze: int):
+        for param in list(self.model.parameters())[:-layers_to_unfreeze]:
+            param.requires_grad = False
+    
+    def train_epoch(self):
```
Comment (vBykoff) [inline Face Recognition/recognizer.py]: аналогично тип возвращаемой переменной

File: Face Recognition/main.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            x = data_transforms[self.mode](x.permute(2, 0, 1))
+            return x
+        else:
+            x, y = self.data[index]
+            x = torch.tensor(x / 255, dtype=torch.float32)
+            x = data_transforms[self.mode](x.permute(2, 0, 1))
+            return x, y
+
+if __name__ == '__main__':
+    root = '../dataset'
```
Comment (vBykoff) [inline Face Recognition/main.py]: было бы неплохо путь передавать через параметры, гугли argparse

File: Face Recognition/main.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        for j in range(5):
+            VP = data['George_W_Bush']
+            idx = np.random.randint(len(VP))
+            image = plt.imread(VP[idx])
+            ax[j].imshow(image)
+            ax[j].set_title('George_W_Bush')
+        plt.tight_layout()
+        plt.show()
+
+    train_faces, val_faces, train_labels, val_labels = train_test_split(faces, labels, test_size=0.3, random_state=42, shuffle=True)
```
Comment (vBykoff) [inline Face Recognition/main.py]: test_size & random state тоже передавать через параметры

File: Face Recognition/main.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+    print(len(train_faces))
+
+    aligned_train_data = [(aligner.get_faces(train_faces[i])[0], train_labels[i]) for i in tqdm(range(len(train_faces)))]
+    aligned_val_data = [(aligner.get_faces(val_faces[i])[0], val_labels[i]) for i in tqdm(range(len(val_faces)))]
+
+    aligned_train_dataset = FacesDataset(aligned_train_data, mode='train')
+    aligned_val_dataset = FacesDataset(aligned_val_data, mode='val')
+
+    for i in range(5):
```
Comment (vBykoff) [inline Face Recognition/main.py]: эти картинки в терминале будет отображаться?

Comment (vBykoff) [reply Face Recognition/main.py]: почему 5, надо бы вынести в отдельную константу заглавными буквами

Comment (vBykoff) [reply Face Recognition/main.py]: картинки можно сохранять в отдельную папку и потом в readme выыводить

File: Face Recognition/main.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        preds = r.predict_one_sample(inputs)
+        predicted = np.argmax(preds)
+        image = plt.imread(test_faces[i])
+        ax[i % 5].imshow(image)
+        ax[i % 5].set_title(label_encoder[predicted])
+        i += 1
+        if i % 5 == 0:
+            plt.tight_layout()
+            plt.show()
+            fig, ax = plt.subplots(nrows=1, ncols=5, figsize=(15, 5))
```
Comment (vBykoff) [inline Face Recognition/main.py:230]: обычно если функция больше 50 строк нужно ее разбивать на подфункции

File: Face Recognition/main.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+            image = plt.imread(test_faces[idx])
+            ax[j].imshow(image)
+        plt.tight_layout()
+        plt.show()
+    
+    aligned_test_data = [aligner.get_faces(test_faces[i])[0] for i in tqdm(range(len(test_faces)))]
+    aligned_test_dataset = FacesDataset(aligned_test_data, mode='test')
+    test_loader = DataLoader(aligned_test_dataset, batch_size=1, shuffle=False)
+
+    i = 0
```
Comment (vBykoff) [inline Face Recognition/main.py]: непонятное название переменной, что оно означает?

File: Face Recognition/main.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+    criterion = nn.CrossEntropyLoss()
+    init_lr = 1e-3
+    optimizer = torch.optim.AdamW(model.parameters(), lr=init_lr)
+
+    batch_size = 32
+    epochs = 50
+    r = recognizer.Recognizer(model, criterion, optimizer, aligned_train_dataset, aligned_val_dataset, batch_size, epochs, device, lr_step=10, lr_coef=0.5)
+
+    layers_to_unfreeze = 30
```
Comment (vBykoff) [inline Face Recognition/main.py]: тоже надо бы вынести в параметры, как и init_lr batch_size epochs

File: Face Recognition/aligner.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
+def draw_boxes(filename: str):
+    import matplotlib.pyplot as plt
```
Comment (vBykoff) [inline Face Recognition/aligner.py]: а зачем импорты внутри функций?

File: Face Recognition/aligner.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+    for i in range(len(faces)):
+        plt.subplot(1, len(faces), i + 1)
+        plt.axis('off')
+        plt.imshow(faces[i])
+    
+    plt.show()
+
+if __name__ == '__main__':
```
Comment (vBykoff) [inline Face Recognition/aligner.py]: это файл со вспомогательными функциями или entrypoint? если 1 зачем тогда main? если 2 то нужны параметры через argparse. И как draw_faces будет отрисовывать в терминале? ТОже надо бы сохранять в отдельную папку если уж надо

File: Face Recognition/main.py [hunk 15]
```diff
@@ -0,0 +1,315 @@
...
+    print("Accuracy: %f" % (test_accuracy))
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(
+        prog="Face recognizer",
+        description="Input path to train dataset, path to test dataset, path to results folder, test size, random state, batch_size, epochs, init_lr, lr_step, lr_coef and layers_to_unfreeze",
+    )
+    parser.add_argument("train_root")
```
Comment (vBykoff) [inline Face Recognition/main.py:242]: можно задавать дефолтные значения если че default = ...

File: Face Recognition/main.py [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+        lr_coef = float(args.lr_coef)
+    else:
+        lr_coef = 1.0
+
+    if args.layers_to_unfreeze:
+        layers_to_unfreeze = int(args.layers_to_unfreeze)
+    else:
+        layers_to_unfreeze = None
+
+    # train_root = '../Dataset/Train'
```
Comment (vBykoff) [inline Face Recognition/main.py:279]: закоментченный код лучше убирать, выведи это в дефолтные параметры argparse

File: README.md [hunk 17]
```diff
@@ -1 +1,23 @@
...
+├── Face Recognition
+│   ├── main.py         <- Main file of the project that launches all computations
+│   │
+│   ├── aligner.py      <- Aligner - a program detecting and cutting faces from the initial photo
+│   │
+│   └── recognizer.py   <- Recognizer - a program recognizing a person in the image that recognizer gets from the aligner
+└── .gitignore
+
+```
+![loss](https://github.com/mkhapisov/Face-recognition/tree/develop/Face%20Recognition/results/loss.jpg)
```
Comment (vBykoff) [inline README.md:22]: че то картинки не отображаются, ну необязятельно в целом, но можно сделать как тут https://github.com/feder-cr/linkedIn_auto_jobs_applier_with_AI

Comment (vBykoff) [reply README.md:22]: там фото вроде 1 в реадми, мб  склонируешь также можешь сделать

**Review by vBykoff — APPROVED** [review]
LGTM
